src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -271,18 +319,76 @@ public void CopyChangedValues(TStructuralType original)
+            // to prevent unrecognizable information from being applied to original resource.
             if (!_entityType.IsAssignableFrom(original.GetType()))
             {
                 throw Error.Argument("original", SRResources.DeltaTypeMismatch, _entityType, original.GetType());
             }
 
-            PropertyAccessor<TStructuralType>[] propertiesToCopy = GetChangedPropertyNames().Select(s => _allProperties[s]).ToArray();
+            RuntimeHelpers.EnsureSufficientExecutionStack();
+
+            // For regular non-structural properties at current level.
+            PropertyAccessor<TStructuralType>[] propertiesToCopy =
+                this._changedProperties.Select(s => _allProperties[s]).ToArray();
             foreach (PropertyAccessor<TStructuralType> propertyToCopy in propertiesToCopy)
             {
                 propertyToCopy.Copy(_instance, original);
             }
 
             CopyChangedDynamicValues(original);
+
+            // For nested resources.
+            foreach (string nestedResourceName in _deltaNestedResources.Keys)
+            {
+                // Patch for each nested resource changed under this TStructuralType.
+                dynamic deltaNestedResource = _deltaNestedResources[nestedResourceName];
+                dynamic originalNestedResource = null;
+                if (!TryGetPropertyRef(original, nestedResourceName, out originalNestedResource))
+                {
+                    throw Error.Argument(nestedResourceName, "Cannot find nested resource name '{0}' in parent resource type '{1}'",
+                        nestedResourceName, original.GetType());
+                }
+
+                if (originalNestedResource == null)
+                {
+                    // When patching original target of null value, directly set nested resource.
+                    dynamic deltaObject = _deltaNestedResources[nestedResourceName];
+                    dynamic instance = deltaObject.GetInstance();
+
+                    // Recursively patch up the instance with the nested resources.
+                    deltaObject.CopyChangedValues(instance);
+
+                    _allProperties[nestedResourceName].SetValue(original, instance);
+                }
+                else
+                {
+                    // Recursively patch the subtree.
+                    deltaNestedResource.CopyChangedValues(originalNestedResource);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the property object by the specified name.
+        /// </summary>
+        /// <param name="structural"></param>
+        /// <param name="propertyName"></param>
+        /// <param name="propertyRef">Output </param>
+        /// <returns></returns>
+        private static bool TryGetPropertyRef(TStructuralType structural, string propertyName,
+            out dynamic propertyRef)
+        {
+            propertyRef = null;
+            bool found = false;
```

Comment:
> bool found = false; [](start = 11, length = 20)

found is necessary?  

line 388, can return "true" directly.
line 391, can return "false" directly

Reply:
Fixed. The only difference is that, as I found out, some ppl prefer single point of return. I am OK with either way.

---
In reply to: [190750787](https://github.com/OData/WebApi/pull/1445#discussion_r190750787) [](ancestors = 190750787)